design/20190708.certificate-request-crd.md

```diff
@@ -0,0 +1,242 @@
+
+TODO: maybe we do want to talk about this as a possibility @munnerz
+
+## Proposal
+
+### API Changes
+
+This proposal will create the following new API types in the
+`certmanager.k8s.io` group;
+
+```
+// CertificateRequestSpec defines the desired state of CertificateRequest
+type CertificateRequestSpec struct {
+	// Requested certificate default Duration
+	// +optional
+	Duration *metav1.Duration `json:"duration,omitempty"`
+
+	// IssuerRef is a reference to the issuer for this CertificateRequest.  If
+	// the 'kind' field is not set, or set to 'Issuer', an Issuer resource with
+	// the given name in the same namespace as the CertificateRequest will be
+	// used.  If the 'kind' field is set to 'ClusterIssuer', a ClusterIssuer with
+	// the provided name will be used. The 'name' field in this stanza is
+	// required at all times. The group field refers to the API group of the
+	// issuer which defaults to 'certmanager.k8s.io' if empty.
+	IssuerRef ObjectReference `json:"issuerRef"`
+
+	// Byte slice containing the PEM encoded CertificateSigningRequest
+	// +optional
+	CSRPEM []byte `json:"csr,omitempty"`
+
+	// IsCA will mark the resulting certificate as valid for signing. This
+	// implies that the 'signing' usage is set
+	// +optional
+	IsCA bool `json:"isCA,omitempty"`
+}
+
+// CertificateStatus defines the observed state of CertificateRequest and
+// resulting signed certificate.
+type CertificateRequestStatus struct {
+	// +optional
+	Conditions []CertificateRequestCondition `json:"conditions,omitempty"`
+
+	// Byte slice containing a PEM encoded signed certificate resulting from the
+	// given certificate signing request.
+	// +optional
+	Certificate []byte `json:"certificate,omitempty"`
+
+	// Byte slice containing the PEM encoded certificate authority of the signed
+	// certificate.
+	// +optional
+	CA []byte `json:"ca,omitempty"`
+}
+```
+
+The `CertificateRequestCondition` resembles much the same of the
+`CertificateRequestCondition`.
+
+The `ObjectReference` field type has had a new field `Group` added as follows:
+
+```
+// ObjectReference is a reference to an object with a given name, kind and group.
+type ObjectReference struct {
+	Name string `json:"name"`
+	// +optional
+	Kind string `json:"kind,omitempty"`
+	// +optional
+	Group string `json:"group,omitempty"`
+}
+```
+
+The group refers to the API group that the target Issuer belongs to. This
+enables namespacing of references to different issuers of external API groups.
+
+### Controller Behaviour
+
+The philociphy for the `CertificateRequest` controllers are planned to be as
+minimal as possible in that the single goal of them is to enable it's owning
+`Issuer` to create the resulting certificate. Once a sync on a
+`CertificateRequest` has been observed, the general flow is as follows:
+
+- Check the group belongs to the owning `Issuer`, exit if not.
+- Check if `CertificateRequest` is in a failed state, exit if true.
```

Comment:
can be followed up later, but we need to more tightly define what `a failed state` is IMO, from an API standpoint